UOP1_Project/Assets/Scripts/DesignPatterns/StateMachine/StateMachine.cs

```diff
@@ -0,0 +1,87 @@
+﻿using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class StateMachine
+{
+	private IState _currentState;
+	
+	private List<Transition> _transitionsForCurrentState = new List<Transition>();
+	private readonly List<Transition> _transitionsFromAnyState = new List<Transition>();
+	
+	private readonly Dictionary<Type, List<Transition>> _allTransitionsForType = new Dictionary<Type,List<Transition>>();
+
+	private static List<Transition> _emptyTransitionsList = new List<Transition>();
+
+	public void Tick()
+	{
+		Transition transition = GetTransitionIfAvailable(_currentState);
+		if (transition != null)
+		{
+			SetState(transition.To);
+		}
+
+		_currentState?.Tick();
+	}
+
+	public void SetState(IState state)
+	{
+		if (state == _currentState)
+		{
+			return;
+		}
+		
+		_currentState?.OnExit();
+		_currentState = state;
+  
+		_allTransitionsForType.TryGetValue(_currentState.GetType(), out _transitionsForCurrentState);
```

Comment:
Could save transitions as part of the states and just call _currentState.transitions() whenever you need them instead.

Reply:
Done and done, quite a good suggestion.